src/main/java/com/faforever/api/user/GogService.java

```diff
@@ -0,0 +1,177 @@
+@Slf4j
+@RequiredArgsConstructor
+public class GogService {
+  // TODO: what characters to allow?
+  private static final Pattern GOG_USERNAME_PATTERN = Pattern.compile("[a-z\\d-_!?.]+"); // gets lower cased by controller, gog doesn't care about capitalization, this causes db entries to be unique
+  private static final Pattern PROFILE_USER_STATUS_PATTERN = Pattern.compile("window\\.profilesData\\.profileUserPreferences\\s=\\s\\{\"bio\":\"(.*?)\"");
+  private static final String GOG_FA_GAME_ID = "1444785261";
+
+  private final FafApiProperties properties;
+  private final Gson gson = new Gson();
+
+  boolean verifyGogUsername(String username) {
+    return username.length() <= 100 && username.length() >= 3 && GOG_USERNAME_PATTERN.matcher(username).matches();
+  }
+
+  boolean verifyProfileToken(String gogUsername, User user, String targetToken) {
+    String profileStatus = getProfileStatus(gogUsername);
+
+    return profileStatus.length() < 100 && Objects.equals(targetToken, profileStatus.trim());
+  }
+
+  private String getProfileStatus(String gogUsername) {
+    String profilePageUrl = String.format(properties.getGog().getProfilePageUrl(), gogUsername);
+    String profilePageHtml = null;
+
+    try {
+      profilePageHtml = new RestTemplate().getForObject(profilePageUrl, String.class);
+    } catch (HttpClientErrorException e) {
+      if (e.getRawStatusCode() == HttpStatus.NOT_FOUND.value()) {
+        throw ApiException.of(ErrorCode.GOG_LINK_PROFILE_NOT_PUBLIC);
+      } else {
+        log.error("Couldn't retrieve gog profile page for {}", profilePageUrl);
+        throw ApiException.of(ErrorCode.GOG_LINK_INTERNAL_SERVER_ERROR);
+      }
+    }
+
+    if (profilePageHtml == null) {
+      log.error("Couldn't retrieve gog profile page for {}", profilePageUrl);
+      throw ApiException.of(ErrorCode.GOG_LINK_INTERNAL_SERVER_ERROR);
+    }
+
+    Document document = Jsoup.parse(profilePageHtml);
+    for (Element element : document.body().getElementsByTag("script")) {
+      String scriptText = element.data();
+      Matcher matcher = PROFILE_USER_STATUS_PATTERN.matcher(scriptText);
+
+      if (matcher.find()) {
+        return matcher.group(1);
+      }
+    }
+
+    throw ApiException.of(ErrorCode.GOG_LINK_PROFILE_NOT_PUBLIC);
+  }
+
+  boolean verifyGameOwnership(String gogUsername) {
+    List<GogGamesListPage.GogGamesListEntry> allGames = getAllGames(gogUsername);
+
+    return allGames.stream().anyMatch(game -> Objects.equals(game.game.id, GOG_FA_GAME_ID));
+  }
+
+  private List<GogGamesListPage.GogGamesListEntry> getAllGames(String gogUsername) {
+    GogGamesListPage firstPage = getGamesListPage(gogUsername, 1);
+
+    List<GogGamesListPage> pages = new ArrayList<>();
+    pages.add(firstPage);
+    for (int i = 2; i <= firstPage.getPages(); i++) {
+      pages.add(getGamesListPage(gogUsername, i));
+    }
+
+    return pages.stream().flatMap(page -> page.get_embedded().getItems().stream()).collect(Collectors.toList());
```

Comment:
This is a bit ugly because you even query all pages if your desired result is on the first page.
But I don't have a class Stream solution for Java 8 at hand.